src/app/qgshandlebadlayers.cpp

```diff
@@ -364,79 +370,48 @@ void QgsHandleBadLayers::editAuthCfg()
 void QgsHandleBadLayers::apply()
 {
   QgsProject::instance()->layerTreeRegistryBridge()->setEnabled( true );
-  QHash<QString, QString> baseChange;
-
 
   for ( int i = 0; i < mLayerList->rowCount(); i++ )
   {
     int idx = mLayerList->item( i, 0 )->data( Qt::UserRole ).toInt();
     QDomNode &node = const_cast<QDomNode &>( mLayers[ idx ] );
 
-    const QString name = mLayerList->item( i, 0 )->text();
     QTableWidgetItem *item = mLayerList->item( i, 4 );
     QString datasource = item->text();
+    QString fileName;
+    const QString name { mLayerList->item( i, 0 )->text() };
     const QString basepath = datasource.left( datasource.lastIndexOf( '/' ) );
-    bool changed = false;
-
-    if ( mFileBase[ name ].size() == 1 )
-    {
-      if ( mFileBase[ name ][0] != basepath && !baseChange.contains( mFileBase[ name ][0] ) )
-      {
-        baseChange[ mFileBase[ name ][0] ] = basepath;
-        changed = true;
-      }
-    }
-    else if ( mFileBase[ name ].size() > 1 )
-    {
-      if ( mFileBase[ name ].indexOf( basepath ) == -1 )
-      {
-        const QList<QString> fileBases = mFileBase[ name ];
-        for ( QString fileBase : fileBases )
-        {
-          if ( !baseChange.contains( fileBase ) )
-          {
-            baseChange[ fileBase ] = basepath;
-            changed = true;
-          }
-        }
-      }
-    }
-    if ( !changed && baseChange.contains( basepath ) )
-      datasource = datasource.replace( basepath, baseChange[basepath] );
+    const QString longName = datasource.mid( datasource.lastIndexOf( '/' ) );
+    if ( longName.lastIndexOf( '|' ) != -1 )
+      fileName = longName.left( longName.lastIndexOf( '|' ) - 1 );
+    else
+      fileName = longName;
+    if ( !( item->foreground() == QBrush( Qt::green ) ) )
```

Comment:
Is there a better way than checking the color? This is neither very readable nor very stable.

Reply:
I could simply add the layer by default but I want the user to validate if the layers that were found are the one that are desired. And not simply layer with the same name but the wrong content.

Otherwise I would have to add a registry if I want user validation.

Reply:
The goal is mostly to have user validation of the found file.

I could either merge it directly and bypass validation but this might lead to wrong input (if the file changed name or if there are duplicates).

Another way would be to maintain a registry of found file. But a more complex behaviour would have to be implemented to ask user for input to approve found layers.

Feedback from more devs would be appreciated on this aspect.

Reply:
You can use `setData` to store the data separate from the representation:

https://doc.qt.io/qt-5/qtablewidgetitem.html#setData

Have a look at e.g. `QgsSettingsTree::Roles` to define custom roles.

Reply:
Couldn't I use Qt::ForegroundRole instead of defining a new role?

Reply:
Better not, that will make it very hard for @nirvn to ship updates for his themes.

Reply:
Right, I'll see what I can come up with.

Reply:
What would be the optimal way to do things, the roles would be associated to the tableitem, now I'm not sure if Qt sets default values for every role. The reason being that if I declare a role in autofind and check for it in apply, there is a possibility that the user will only go for apply and if the desires role has not been set or has the wrong default value, this could have unintended consequences.

I'm not too sure if fiddling with roles and potentially uninitiated values is that much of a good idea.

I would assume that the color is carried as it can be assigned to a role. Since this role/foreground is only bound to the table item I'm missing the point as how this would affect the theme of qgis.  By using a default display role I have less of a risk to stumble upon a wrong value when checking the data.

Reply:
set user data via setData is indeed the correct approach for something like this. The danger with using the colour is that it mixes presentation attributes with non-presentation, and it's almost certainly going to break in future when someone comes along and tweaks the shade of green used (not realising that you're relying on an exact colour check for other stuff elsewhere).

There's zero risk of uninitialised values. If the user data hasn't been set, you'll get an invalid QVariant value returned for it. These can be tested for by checking QVariant::isValid()

I'd suggest digging into qts docs on QVariant and the model/view approach. These are used everywhere throughout QGIS, so it's handy to know them well! 😀

Reply:
I have implemented something, using UserRole.